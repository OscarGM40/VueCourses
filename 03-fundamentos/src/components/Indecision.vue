<template>
  <h1>Indecision</h1>
  <!-- <img v-bind:src="img" alt="bg" /> -->
  <img v-show="img" :src="img" alt="bg" />
  <div class="bg-dark"></div>
  <div class="indecision-container">
    <!--  FORMA CON EVENTO KEYPRESS @keypress.enter="handleInput" -->
    <input
      type="text"
      id="question"
      v-model="question"
      placeholder="Hazme una pregunta"
    />
    <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>
    <div v-if="isValidQuestion">
      <h2>{{ this.question }}</h2>
      <h1>{{ answer }}</h1>
    </div>
  </div>
</template>

<script>
export default {
  data() {
    return {
      question: "",
      answer: null,
      img: null,
      isValidQuestion: false,
    };
  },
  methods: {
    handleInput() {
      this.question = document.getElementById("question").value;
    },
    async getAnswer() {
      const responses = {
        yes:"Si!",
        no:"No!",
        maybe: "Quizas!"
      }
      this.answer = "Pensando...";
      const { answer, image } = await fetch("https://yesno.wtf/api").then((r) =>
        r.json()
      );
      this.answer = responses[answer];
      this.img = image;
    },
  },
  watch: {
    question(value, oldValue) {
      this.isValidQuestion = false;
      if (!value.includes("?")) return;

      this.isValidQuestion = true;
      this.getAnswer();
    },
  },
  name: "Indecision",
};
</script>

<style scoped>
img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  position: relative;
  z-index: 99;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: none;
}
input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 10px;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>
