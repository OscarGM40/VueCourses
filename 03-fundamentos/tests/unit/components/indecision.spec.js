import { shallowMount } from "@vue/test-utils";
import Indecision from "@/components/Indecision";

describe("Indecision Component", () => {
  let wrapper;
  // un spy es un espia que está pendiente de cualquier cosa,pueden ser métodos,pueden ser librerias completas(modulos),
  let clgSpy;
  // dado que un test se ejecuta sobre Node no va a ver la api fetch
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          answer: "yes",
          forced: false,
          image: "https://image.com",
        }),
    }),
  );

  beforeEach(() => {
    wrapper = shallowMount(Indecision);
    clgSpy = jest.spyOn(console, "log");
    jest.clearAllMocks();
  });

  test("Debe de hacer match contra el snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("no debe de disparar la petición si no escribo en el input el simbolo de interrogacion", async () => {
    const input = wrapper.find("input"); //solo hay uno,asinto
    // dado que escribir en el input es asincrono tengo que usar await(fijate que Vue es muy claro en cuanto a que me indica exactamente la acción asincrona,me será útil)
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    console.log(wrapper.vm);

    await input.setValue("Hola mundo");
    // 2 porque cuenta el de testing de la linea 24 tmb(fijate que es importante pues)
    expect(clgSpy).toHaveBeenCalledTimes(2);
    expect(getAnswerSpy).not.toHaveBeenCalled();
  });

  test("escribir la interrogacion si debe disparar la peticion", async () => {
    const input = wrapper.find("input");
    // fijate que este test no necesita un espia,me da igual lo que haga el método
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    const fetch = jest.spyOn(window, "fetch");
    await input.setValue("Hola mundo?");
    expect(getAnswerSpy).toHaveBeenCalled();
  });

  test("pruebas en el metodo getAnswer- Peticion exitosa", async () => {
    // llamamos fisicamente al método(fijate que es asincrono)
    await wrapper.vm.getAnswer();
    console.log(wrapper.vm.img);
    console.log(wrapper.vm.answer);
    expect(wrapper.vm.img).toBe("https://image.com");
    expect(wrapper.vm.answer).toBe("Si!");
  });
  test("pruebas en el metodo getAnswer- La API debe fallar", async () => {
    // fijate que ya tengo definido el mock sobre la propiedad global.fetch del browser en la linea 9,luego ahora ya no es jest.fn() sino que es un update usando mockImplementation
    fetch.mockImplementationOnce(() => Promise.reject("API is down")); // fijate que nos da igual usar global.fetch o simplemente fetch

    // llamamos al método
    await wrapper.vm.getAnswer();
    // también podemos llamar al espia,que va a hacer la llamada física igualmente con la implementación real
    // const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    // getAnswerSpy();

    const img = wrapper.find("img");
    // truthy porque usé v-show,falsy si fuera v-if
    expect(img.exists()).toBeTruthy();
    expect(wrapper.vm.answer).toBe("No se pudo cargar del API");
  });
});
