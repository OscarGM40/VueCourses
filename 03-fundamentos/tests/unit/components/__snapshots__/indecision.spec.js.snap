// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component Debe de hacer match contra el snapshot 1`] = `
<h1>Indecision</h1>
<!-- <img v-bind:src="img" alt="bg" /> -->
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container">
  <!--  FORMA CON EVENTO KEYPRESS @keypress.enter="handleInput" --><input type="text" id="question" placeholder="Hazme una pregunta">
  <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>
  <!--v-if-->
</div>
`;
